ext/net/tcp.rs

```diff
@@ -0,0 +1,168 @@
+/// load balancing among the sockets, whereas macOS follows a "last bind wins" strategy.
+pub struct TcpLbListener {
+  listener: Option<tokio::net::TcpListener>,
+  conn: Option<Arc<TcpConnection>>,
+}
+
+impl TcpLbListener {
+  /// Bind to a port. If `reuse_port` is specified and this is not Linux, we use a shared
+  /// process-wide port.
+  pub fn bind(
+    socket_addr: SocketAddr,
+    reuse_port: bool,
+  ) -> std::io::Result<Self> {
+    if cfg!(not(target_os = "linux")) && reuse_port {
+      Self::bind_load_balanced(socket_addr)
+    } else {
+      Self::bind_direct(socket_addr, reuse_port)
+    }
+  }
+
+  /// Bind directly to the port.
+  fn bind_direct(
+    socket_addr: SocketAddr,
+    reuse_port: bool,
+  ) -> std::io::Result<Self> {
+    let listener = bind_socket_and_listen(socket_addr, reuse_port)?;
+    Ok(Self {
+      listener: Some(tokio::net::TcpListener::from_std(listener)?),
+      conn: None,
+    })
+  }
+
+  /// Bind to the port in a load-balanced manner.
+  fn bind_load_balanced(socket_addr: SocketAddr) -> std::io::Result<Self> {
+    let tcp = &mut CONNS.get_or_init(Default::default).lock().unwrap().tcp;
+    if let Some(conn) = tcp.get(&socket_addr) {
+      let listener = Some(conn.listener()?);
+      return Ok(Self {
+        listener,
+        conn: Some(conn.clone()),
+      });
+    }
+    let conn = Arc::new(TcpConnection::start(socket_addr)?);
+    let listener = Some(conn.listener()?);
+    tcp.insert(socket_addr, conn.clone());
+    Ok(Self {
+      listener,
+      conn: Some(conn),
+    })
+  }
+
+  pub async fn accept(
+    &self,
+  ) -> std::io::Result<(tokio::net::TcpStream, SocketAddr)> {
+    let (tcp, addr) = self.listener.as_ref().unwrap().accept().await?;
+    Ok((tcp, addr))
+  }
+
+  pub fn local_addr(&self) -> std::io::Result<SocketAddr> {
+    self.listener.as_ref().unwrap().local_addr()
+  }
+}
+
+impl Drop for TcpLbListener {
+  fn drop(&mut self) {
+    // If we're in load-balancing mode
+    if let Some(conn) = self.conn.take() {
+      let mut tcp = CONNS.get().unwrap().lock().unwrap();
+      if Arc::strong_count(&conn) == 2 {
+        tcp.tcp.remove(&conn.key);
+        // Close the connection
+        debug_assert_eq!(Arc::strong_count(&conn), 1);
+        drop(conn);
+      }
+    }
+  }
+}
+
+/// Bind a socket to an address and listen with the low-level options we need.
+#[allow(unused_variables)]
+fn bind_socket_and_listen(
+  socket_addr: SocketAddr,
+  reuse_port: bool,
+) -> Result<std::net::TcpListener, std::io::Error> {
+  let socket = if socket_addr.is_ipv4() {
+    socket2::Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))?
+  } else {
+    socket2::Socket::new(Domain::IPV6, Type::STREAM, Some(Protocol::TCP))?
+  };
+  #[cfg(target_os = "linux")]
+  if reuse_port {
+    socket.set_reuse_port(true)?;
+  }
+  #[cfg(not(windows))]
+  socket.set_reuse_address(true)?;
```

Comment:
This existed before and prevents most of our tests from failing when we reuse a port binding in a subsequent test.